### ISSUE TITLE
alternate way of setup ha with master-slave mode

### ISSUE THREAD

**Issue by hanbeibei** [jupyterhub/jupyterhub#2068]
I want to setup ha, and got this: Support HA (High Availability) #1932

seems we setup multiple hub is not available,  let's make it happened in an alternative way: master slave mode.  if the master hub crashed, the slave hub will take over all the docker container.

I tried to install and configure 2 hubs, but if i kill the master hub's process id, the slave hub can not take over the all the docker containers.

sorry I can not post the log, they are all located in company intranet.

the slave hub continuous report the 302 error. and the container report: failed to establish a new connection: connection refulsed.

Could you help me to point out the obstacle?

Great appreciate

Great appreciate 
Oliver

**Comment by minrk** [comment]
Yes, JupyterHub cannot run in a hot-swappable 'failover' mode. JupyterHub has in-memory state, so two running JupyterHubs pointed to the same database will diverge in their understanding of what's running. It only reconstructs its state from the database on startup. A new JupyterHub instance can take over from a previous one that died, but only if it starts after the previous one stopped.

If you want to have a 'hot' container ready to swap-in for another one so your delay is only process startup and not container creation, then the container should hold in a state prior to starting the jupyterhub application, and only start the hub when the previous instance is known to be gone. The shortest window version of this would be a Python process that has imported `jupyterhub.app` and calls `jupyterhub.app.main()` when it's time to take over.

**Comment by hanbeibei** [comment]
thanks @minrk  for your kindly reply.

maybe I need to show more detail, in nginx.conf, we setup two jupyterhub servers:

10.100.16.57
10.100.16.58, backup

and the two jupyterhub are running, nginx will connect 10.100.16.57, if I kill the process id on 10.100.16.57, 10.100.16.58 can not take over all the running container.

got the error on 10.100.16.58: 302, it's redirection and on container: websocket disconnect and then unable to establish a new connection, 

so the slave hub is always on and is ready for take over if master hub crash.

Great appreciate.
Oliver

**Comment by hanbeibei** [comment]
ah... problem resolved, it is the slave hub can not recognize the container, add nginx's .pem file, it works fine..

**Comment by hanbeibei** [comment]
@willingc this issue can be closed, thanks again!
